torch/testing/_asserts.py

```diff
@@ -370,194 +370,181 @@ def _amend_error_message(exc: E, msg_fmtstr: str) -> E:
-    elif isinstance(actual, collections.abc.Mapping) and isinstance(expected, collections.abc.Mapping):
-        return _parse_mappings(actual, expected)
-    else:
-        return _parse_array_or_scalar_like_pair(actual, expected)
+    else:  # isinstance(pair, TensorPair)
+        return check_tensors(pair.actual, pair.expected)
 
 
-def _parse_array_or_scalar_like_pair(actual: Any, expected: Any) -> Tuple[Optional[Exception], Optional[_ParsedInputs]]:
-    """Parses an scalar-or-array-like pair.
+def _to_tensor(array_or_scalar_like: Any) -> Tuple[Optional[Exception], Optional[Tensor]]:
+    """Converts a scalar-or-array-like to a :class:`~torch.Tensor`.
 
     Args:
-        actual: Actual array-or-scalar-like.
-        expected: Expected array-or-scalar-like.
+        array_or_scalar_like (Any): Scalar-or-array-like.
 
     Returns:
-        (Optional[Exception], Optional[_ParsedInputs]): The two elements are orthogonal, i.e. if the first ``is None``
-            the second will not and vice versa. Returns a :class:`UsageError` if :attr:`actual` and :attr:`expected` do
-            not have the same type or no :class:`~torch.Tensor` can be constructed from them.
+        (Tuple[Optional[Exception], Optional[_TensorPair]]): The two elements are orthogonal, i.e. if the first
+            ``is None`` the second will be valid and vice versa. Returns a :class:`UsageError` if no tensor can be
+            constructed from :attr:`actual` or :attr:`expected`. Additionally, returns any exception from
+            :func:`_check_supported_tensor`.
     """
     exc: Optional[Exception]
 
-    if type(actual) is not type(expected):
-        exc = UsageError(
-            f"Apart from a containers type equality is required, but got {type(actual)} and {type(expected)} instead."
-        )
-        return exc, None
-
-    tensors = []
-    for array_or_scalar_like in (actual, expected):
+    if isinstance(array_or_scalar_like, Tensor):
+        tensor = array_or_scalar_like
+    else:
         try:
             tensor = torch.as_tensor(array_or_scalar_like)
         except Exception:
             exc = UsageError(f"No tensor can be constructed from type {type(array_or_scalar_like)}.")
             return exc, None
 
-        exc = _check_supported_tensor(tensor)
-        if exc:
-            return exc, None
-
-        tensors.append(tensor)
-
-    actual_tensor, expected_tensor = tensors
-    return None, _ParsedInputs(actual_tensor, expected_tensor)
+    exc = _check_supported_tensor(tensor)
+    if exc:
+        return exc, None
 
+    return None, tensor
 
-def _parse_sequences(actual: Sequence, expected: Sequence) -> Tuple[Optional[Exception], Optional[_ParsedInputs]]:
-    """Parses sequences of scalar-or-array-like pairs.
 
-    Regardless of the input types, the sequences are returned as :class:`list`.
+def _to_tensor_pair(actual: Any, expected: Any) -> Tuple[Optional[Exception], Optional[_TensorPair]]:
```

Comment:
It does seem like there's a lot of architecture churn to support this scenario, we probably want to review the architecture holistically before 1.9, if possible

Reply:
We already had this before: `_parse_array_or_scalar_like_pair` on the old L443. We always need to parse a pair to give expressive messages for offending index / key in case of containers.